integration-tests/test/rnode.py

```diff
@@ -334,6 +335,30 @@ def log_lines(self) -> List[str]:
         log_content = self.logs()
         return Node.__log_message_rx.split(log_content)
 
+    def deploy_contract_with_substitution(self, substitute_dict: Dict[str, str], rho_file_path: str, private_key: str) -> str:
+        # Supposed that you have a contract with content like below.
+        #
+        # new x in { x!("#DATA") }
+        #
+        # If you pass a dict {'#DATA': "123456"} as substitute_dict args in this func,
+        # this method would substitute the string #DATA in the contract with 123456, which turns out to be
+        #
+        # new x in { x!("123456") }
+        #
+        # And then deploy the contract in the node
+        shutil.copyfile(rho_file_path, os.path.join(self.local_deploy_dir, os.path.basename(rho_file_path)))
+        container_contract_file_path = os.path.join(self.remote_deploy_dir, os.path.basename(rho_file_path))
+        substitute_rules = ';'.join(['s/{}/{}/g'.format(key, value) for key, value in substitute_dict.items()])
+        self.shell_out(
```

Comment:
`/` should be probably escaped in `key` and `value`.

Reply:
Why has this been resolved?

Reply:
`'s/{}/{}/g'` -> `r's/{}/{}/g'`
Actually `/` doesn't have to be escaped while `\` should be escaped.
I put a `r` in front of the string which means this is the raw string and even the `\` doesn't have to be escaped.

Reply:
https://stackoverflow.com/questions/2081640/what-exactly-do-u-and-r-string-flags-do-and-what-are-raw-string-literals

Reply:
Try to run
```
x=/
echo /foo/bar | sed "s/$x/;/"
```
You'll get an error. You can use different delimiters in sed, e.g. `|` or `#` instead of `/` to avoid this.

Reply:
What about `x=#`? I don't think that's the way to solve the error.

Reply:
`/` as delimiters are fine.  It's just that they need to be escaped or else sed will fail on cases like `'s/{}/{}/g'.format('foo/bar', 'anything')`

Reply:
https://docs.python.org/3/library/shlex.html#shlex.quote should work better than str.replace